src/ide/tgui.ts

```diff
@@ -202,6 +212,27 @@ export let tgui = (function () {
 		};
 	};
 
+	function setupCanvasZoom(canvas) {
+		// zoom
+		// TODO: programmaticly detect whenever zoom changes
+		let zoom = 2; // Good enough
+		//let zoom = window.devicePixelRatio;
+		if (zoom > 1) {
+			canvas.width *= zoom;
+			canvas.height *= zoom;
+			let ctx = canvas.getContext("2d");
+			ctx.scale(zoom, zoom);
+		}
+	}
+
+	function renderCanvas(canvas, draw) {
+		if (typeof draw === "string") {
+			console.assert(draw.substring(0, 5) == "icon:");
+			let id = draw.substring(5);
+			icons[id](canvas, module.dark_theme);
```

Comment:
What I dislike here is the use of strings to identify the icons. I think it would be better if you left the drawing functions in the description object and used a [`WeakMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) to assign the drawing functions their canvas elements. When changing the theme you could then use that `WeakMap` again to get the drawing functions belonging to each canvas.

Another, more self-contained approach might be to emit a custom event on `body` when the theme changes with each canvas listening to it. This might be more complex however...

Reply:
I have used strings here, because those can be stored in the dom. I think about changing it in such a way, that strings might be used under the hood, but when assigning an icon, the elements in the icons module are accessed.

Reply:
Again, you don't need to move the drawing functions to a new file and use strings. If you used a `WeakMap` to store the drawing functions, you could just keep them where they were and change much less.

#### Example
You create a `WeakMap` to store the functions:
```ts
type IconDrawingFunction = (canvas: HTMLCanvasElement, dark: boolean) => void;
const iconDrawingFns = new WeakMap<HTMLCanvasElement, IconDrawingFunction>();
```

You store the drawing function for the canvas:
```ts
iconDrawingFns.set(canvas, draw);
draw(canvas, isDarkTheme(module.theme));
```

And when the theme changes, you can just get them from the `WeakMap` again:
```ts
const draw = iconDrawingFns.get(canvas);
draw(canvas, isDarkTheme(newTheme));
```

Doing it this way has a much smaller impact and allows the icon drawing functions to stay where they are.

In addition, it might be better to keep them where they are. You can currently see whether and where they are used, because they are directly referenced and mostly even declared inline. When moving them away it gets a little harder to tell where a specific icon is used and you need to remember removing one once it's no longer used.

Reply:
Again, you don't need to move the drawing functions to a new file and use strings. If you used a `WeakMap` to store the drawing functions, you could just keep them where they were and change much less.

#### Example
You create a `WeakMap` to store the functions:
```ts
type IconDrawingFunction = (canvas: HTMLCanvasElement, dark: boolean) => void;
const iconDrawingFns = new WeakMap<HTMLCanvasElement, IconDrawingFunction>();
```

You store the drawing function for the canvas:
```ts
iconDrawingFns.set(canvas, draw);
draw(canvas, isDarkTheme(module.theme));
```

And when the theme changes, you can just get them from the `WeakMap` again:
```ts
const draw = iconDrawingFns.get(canvas);
draw(canvas, isDarkTheme(newTheme));
```

Doing it this way has a much smaller impact and allows the icon drawing functions to stay where they are.

In addition, it might be better to keep them where they are. You can currently see whether and where they are used, because they are directly referenced and mostly even declared inline. When moving them away it gets a little harder to tell where a specific icon is used and you need to remember removing one once it's no longer used.

Reply:
I suggest that the icons by themselves should be in a separate file, as I have currently implemented it here. This way the ide/gui source code is not cluttered as much. But I see the advantage of inline icons. Also when adding a new toolbar button, only one file needs to be changed.